Modify the definition of the contact frame - make it more robust

SeanCurtis-TRI opened the pull request:
The old contact frame suffered from robustness issues.  Normals that were
close to the z-axis would lead to robustness issues in the required
cross products.

This approach maximizes the perpendicularity of the vectors that are
being crossed leading to a more generally robust calculation.

Based on issue #4126

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4272)
<!-- Reviewable:end -->

SeanCurtis-TRI commented:
Simple patch: +@sherm1 for platform and feature review.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXg5jt7LTD2z2weW93f:bro1swd)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
This is great! But don't worry, no good deed goes unpunished! Reviewed -- PTAL.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 458 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgzRHcFWeIDF2C1teh:-KXgzRHcFWeIDF2C1tei:brvggch) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L458)):*
> ```cpp
>       Vector3<T> this_normal = pair.normal;
> 
>       // Creates a local basis for contact.  The frame's z-axis is the normal
> ```

The contact frame should have a symbol so it is easier to talk about below. I suggest "C"!

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 460 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXh1Z-Q7-WRRFCHzRol:-KXh1Z-Q7-WRRFCHzRom:bwxs3fc) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L460)):*
> ```cpp
>       // Creates a local basis for contact.  The frame's z-axis is the normal
>       // direction of the contact.  The other two vectors are arbitrary
>       // vectors that form the orthonormal basis.
> ```

Sean, can you put this frame-maker into a separate method of RBP and add a unit test for it?

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 461 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgz762jTJeOsuJYM24:-KXgz762jTJeOsuJYM25:bmm39le) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L461)):*
> ```cpp
>       // direction of the contact.  The other two vectors are arbitrary
>       // vectors that form the orthonormal basis.
>       Vector3<T> tangent1;
> ```

BTW, better to call this `x_axis` I think.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 476 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgzEUmC-G26X87ouIS:-KXgzEUmC-G26X87ouIT:b5tb6wl) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L476)):*
> ```cpp
>       tangent1 = this_normal.cross(perpAxis).normalized();
> 
>       Vector3<T> tangent2 = this_normal.cross(tangent1);
> ```

BTW, `y_axis=z_axis.cross(x_axis)` would be easier to understand.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 478 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgydaCbBvfFQdfrR1P:-KXgydaCbBvfFQdfrR1Q:bi07o85) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L478)):*
> ```cpp
>       Vector3<T> tangent2 = this_normal.cross(tangent1);
>       // Transformation from world frame to local surface frame.
>       Matrix3<T> R;
> ```

This is ambiguous as written. Because frame C is being written into the *rows* here, I believe this is R_CW (=R_WC transpose). Please rename (to whatever is the correct interpretation). R_CW makes sense if JA and JB are world-frame and we are trying to calculate the relative velocity in the C frame. 

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 483 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXh-WfOS00d88hX5qkb:-KXh-WfOS00d88hX5qkc:bifed6k) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L483)):*
> ```cpp
>       R.row(2) = this_normal;
>       auto J = R * (JA - JB);  // J = [ D1; D2; n ]
>       // [ tangent1dot; tangent2dot; phidot ]
> ```

I don't understand this comment. This is just the relative translational velocity between the point on A and the point on B, not the time derivative of the axes. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXh1low5OPlaamOiiR3:b-qymccl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Thanks.  :)  One day I'll stop assuming quick and dirty will ever win.  :)  I've moved some code around (as per request). Your old comments may not map to the new code very well.  PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 458 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgzRHcFWeIDF2C1teh:-KXkvuIQbAGQd-x1Seh4:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L458)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The contact frame should have a symbol so it is easier to talk about below. I suggest "C"!
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 460 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXh1Z-Q7-WRRFCHzRol:-KXkvwkNsYdwvuYsWgXr:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L460)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Sean, can you put this frame-maker into a separate method of RBP and add a unit test for it?
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 461 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgz762jTJeOsuJYM24:-KXh9rGXuu0d4F58yYsE:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L461)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, better to call this `x_axis` I think.
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 476 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgzEUmC-G26X87ouIS:-KXkvyPodVHiMoA3bfKD:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L476)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, `y_axis=z_axis.cross(x_axis)` would be easier to understand.
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 478 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXgydaCbBvfFQdfrR1P:-KXkvzdVwrMZAyWhwRf_:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L478)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This is ambiguous as written. Because frame C is being written into the *rows* here, I believe this is R_CW (=R_WC transpose). Please rename (to whatever is the correct interpretation). R_CW makes sense if JA and JB are world-frame and we are trying to calculate the relative velocity in the C frame. 
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 483 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXh-WfOS00d88hX5qkb:-KXh4P5ulz8KNUCH02hC:b3wg65q) ([raw file](https://github.com/robotlocomotion/drake/blob/7a193462ce8ecbf1db653422019b8be4fec25fba/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L483)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I don't understand this comment. This is just the relative translational velocity between the point on A and the point on B, not the time derivative of the axes. 
</blockquote></details>

Done

No peeking outside of the scope of this PR.  That's a legacy comment.  However, I've removed it as I couldn't account for it either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXh3td06xRxkBLQU0Ns:b-oz5v1y)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 481 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXkyXdp0k8Tb2-yB6MJ:-KXkyXdp0k8Tb2-yB6MK:b-md0tzy) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L481)):*
> ```cpp
>       Vector3<T> this_normal = pair.normal;
> 
>       // rotation from world (W) to contact frame (C), e.g., q_C = R_CW * q_W.
> ```

Are there missing words before "rotation"? Maybe "Defines the"?

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 438 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXky6e11mqfzvGW_bru:-KXky6e11mqfzvGW_brv:bixdxlg) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L438)):*
> ```cpp
> }
> 
> // This tests the contact frame to confirm that a robust orthonormal frame
> ```

BTW: nit: Consider replacing "This tests" with "Tests" to be consistent with preceding​ unit test's comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 481 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXkyXdp0k8Tb2-yB6MJ:-KXl9qaX_EPWOMAYcMzH:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L481)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Are there missing words before "rotation"? Maybe "Defines the"?
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 438 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXky6e11mqfzvGW_bru:-KXl9rfwioVSz2gJCKm_:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L438)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: Consider replacing "This tests" with "Tests" to be consistent with preceding​ unit test's comment.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 199 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlA6Ic10Xm106mfkDV:-KXlA6Ic10Xm106mfkDW:bp0mb8t) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.h#L199)):*
> ```objc
>   /// the frame in which the z-axis is expressed and L is a local basis such
>   /// that q_L = R_LW * q_W.
>   /// @param[in]  z_axis    The vector defining the basis's z-axis.
> ```

" ... and expressed in a frame W."

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 201 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXl7c6UgfG1mOj1alml:-KXl7c6UgfG1mOj1almm:b-hn8pll) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.h#L201)):*
> ```objc
>   /// @param[in]  z_axis    The vector defining the basis's z-axis.
>   /// @param[out] R_LW      The final basis.
>   void ComputeBasisFromZ(const Vector3<T>& z_axis, Matrix3<T>* R_LW) const;
> ```

Great - thanks! May I suggest a few refinements:
- Return the basis/rotation matrix as the function return rather than as an output argument (that's cost-free in C++11 and nicer to use). You can use doxygen's `@retval` to give the return argument a meaningful name.
- Consider more suggestive symbols, "F" for "frame" (rather than L) and "v" for vector (rather than q).
- Consider returning the more conventional R_WF rather than the transpose (that is, fill in R with columns (x;y;z) rather than rows. Eigen's optimizations allow the matrix and its transpose to be used equally efficiently. The contact code uses the rotation both ways anyway, and I would rather see us use this convention everywhere.
- Consider being even more excruciatingly precise about frames. I would write `z_W` or `z_axis_W` just to make sure everyone gets the message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks for these changes. Fresh comments, PTAL.

---

Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 450 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlFD9pJN5zeL2bW5GI:-KXlFD9pJN5zeL2bW5GJ:b1cvklq) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L450)):*
> ```cpp
>   x_axis = z_axis.cross(perpAxis).normalized();
>   Vector3<T> y_axis = z_axis.cross(x_axis);
>   // Transformation from world frame to local frame.
> ```

BTW, oh that's why you called it "L"!

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 477 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlE-F3RtRXTiCwGFRd:-KXlE-F3RtRXTiCwGFRe:b-m5jg9c) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L477)):*
> ```cpp
>       auto JA = tree_->transformPointsJacobian(kinsol, pair.ptA, body_a_index,
>                                                0, false);
>       auto JB = tree_->transformPointsJacobian(kinsol, pair.ptB, body_b_index,
> ```

BTW, now that you have sorted out all the frames consider enshrining that in the symbol names. In particular, `JA_W` and `JB_W` and `J_C` below would go a long way to making this code understandable.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 483 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlAwIGaFJCxFA0DO_z:-KXlAwIGaFJCxFA0DOa-:b-ubbzdc) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L483)):*
> ```cpp
>       // R_CW is a rotation from world (W) to contact frame (C),
>       // e.g., q_C = R_CW * q_W.
>       Matrix3<T> R_CW;
> ```

BTW, I suggest using `v` for generic vector rather than `q`.
Also, the "from ... to" language is ambiguous without the equation so consider
something like "Rotation matrix R_CW takes vector v_W expressed in the world frame W and re-expresses it in C when used on the left: v_C = R_CW*v_W."

(It works the opposite way on the right: v_W'=vC'*R_CW where the primes mean transpose.)

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 484 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlEZ69lTbIZA4YhUOT:-KXlEZ69lTbIZA4YhUOU:b5mg5r3) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L484)):*
> ```cpp
>       // e.g., q_C = R_CW * q_W.
>       Matrix3<T> R_CW;
>       ComputeBasisFromZ(this_normal, &R_CW);
> ```

BTW, if you liked my earlier suggestions this would be:
`const Matrix3<T> R_WC = ComputeBasisFromZ(this_normal);` (or maybe `this_normal_W`?)

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 201 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlG8Ylpp-HZ3yQLOTC:-KXlG8Ylpp-HZ3yQLOTD:bjej8j2) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.h#L201)):*
> ```objc
>   /// @param[in]  z_axis    The vector defining the basis's z-axis.
>   /// @param[out] R_LW      The final basis.
>   void ComputeBasisFromZ(const Vector3<T>& z_axis, Matrix3<T>* R_LW) const;
> ```

This method should be static.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 436 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlFfM2CE2PWNsjNNOL:-KXlFfM2CE2PWNsjNNOM:b-yhcnkj) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L436)):*
> ```cpp
>   EXPECT_NEAR(0.0, R.col(0).dot(R.col(2)), kEpsilon);
>   EXPECT_NEAR(0.0, R.col(1).dot(R.col(2)), kEpsilon);
> }
> ```

Add a check for right-handedness (col2 = col0 X col1).

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 444 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlGDj6e13RPx4fXc6L:-KXlGDj6e13RPx4fXc6M:bjzu9k) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L444)):*
> ```cpp
>   // being tested does not actually require a valid RigidBodyTree.  So, this
>   // uninitialized tree is sufficient.
>   RigidBodyPlant<double> plant(make_unique<RigidBodyTree<double>>());
> ```

The method under test should be static since it doesn't need access to RBPlant's data members. Then you can invoke it `RigidBodyPlant<double>::ComputeBasisFromZ()` and don't need a runtime object.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXlGknIJFwBoSwV6k1u:bre0tyu)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Comments addressed. PTAL

---

Review status: 0 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 450 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlFD9pJN5zeL2bW5GI:-KXlL0omGzWe8s3RNMdZ:b-hzhoyy) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L450)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, oh that's why you called it "L"!
</blockquote></details>

BTW I've tweaked the documentation to make "L" more obvious.  It no longer creates a basis, it creates a local basis.  Hopefully that'll help.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 477 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlE-F3RtRXTiCwGFRd:-KXlL9dQkF7M3Ql5KpEJ:bzeaf0r) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L477)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, now that you have sorted out all the frames consider enshrining that in the symbol names. In particular, `JA_W` and `JB_W` and `J_C` below would go a long way to making this code understandable.
</blockquote></details>

BTW I'm going to skip this.  First, its outside the scope of this PR. Second, the contact model code still needs to be gone over more generally.  I think this is polishing the rock when what we really need is a new rock. :)

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 483 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlAwIGaFJCxFA0DO_z:-KXlMF4htlZvlNMcNcDt:b-llew9n) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L483)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I suggest using `v` for generic vector rather than `q`.
Also, the "from ... to" language is ambiguous without the equation so consider
something like "Rotation matrix R_CW takes vector v_W expressed in the world frame W and re-expresses it in C when used on the left: v_C = R_CW*v_W."

(It works the opposite way on the right: v_W'=vC'*R_CW where the primes mean transpose.)
</blockquote></details>

BTW I'd say that "from ... to" language is no more ambiguous as what you've written.  The magical phrase is "when used on the left."  Change that to "used on the right" and the whole rest of your description must likewise change.  

That said, the doc can be more precise and I've done so.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 484 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlEZ69lTbIZA4YhUOT:-KXlMv7XkmIwFYv0JFIA:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L484)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, if you liked my earlier suggestions this would be:
`const Matrix3<T> R_WC = ComputeBasisFromZ(this_normal);` (or maybe `this_normal_W`?)
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 199 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlA6Ic10Xm106mfkDV:-KXlC7t2aSobpbuTZmmy:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.h#L199)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

" ... and expressed in a frame W."
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 201 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXl7c6UgfG1mOj1alml:-KXlC6eFaLG0A9I1oI87:b-17ylvh) ([raw file](https://github.com/robotlocomotion/drake/blob/b9a366be687c0f051b59e51c9d4408034b136674/drake/multibody/rigid_body_plant/rigid_body_plant.h#L201)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Great - thanks! May I suggest a few refinements:
- Return the basis/rotation matrix as the function return rather than as an output argument (that's cost-free in C++11 and nicer to use). You can use doxygen's `@retval` to give the return argument a meaningful name.
- Consider more suggestive symbols, "F" for "frame" (rather than L) and "v" for vector (rather than q).
- Consider returning the more conventional R_WF rather than the transpose (that is, fill in R with columns (x;y;z) rather than rows. Eigen's optimizations allow the matrix and its transpose to be used equally efficiently. The contact code uses the rotation both ways anyway, and I would rather see us use this convention everywhere.
- Consider being even more excruciatingly precise about frames. I would write `z_W` or `z_axis_W` just to make sure everyone gets the message.
</blockquote></details>

Done

   - Done
   - For me this comes down to type vs meaning.  "L" and "W" refer to local frame and world frame, respectively. Both frames whose interpretations are built into the name.  Whereas "F" refers to frame, just a type. So, I'm inclined to leave it as L.  But I have swapped `q` for `v`.
   - Done. I wrestled on this one.  Changing the matrix means making sure I don't introduce errors; but I agree that concatenating column vectors is the more traditional means of constructing a basis matrix.  So, at the end of the day, I chose to make this change.
   - Done

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 201 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlG8Ylpp-HZ3yQLOTC:-KXlKiBt4X2JLn3Yov_S:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/rigid_body_plant.h#L201)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This method should be static.
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 436 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlFfM2CE2PWNsjNNOL:-KXlKjPkVoxmONECiR1o:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L436)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Add a check for right-handedness (col2 = col0 X col1).
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 444 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlGDj6e13RPx4fXc6L:-KXlKvkax6zdZJikZD3j:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/2a62650bb1a387af1ee0cc5e2ef5cc5de511223d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L444)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The method under test should be static since it doesn't need access to RBPlant's data members. Then you can invoke it `RigidBodyPlant<double>::ComputeBasisFromZ()` and don't need a runtime object.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXlKxg48cubg2dxCHt5:byq25nx)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Two non-blocking comments (one you should fix). Then <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 439 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlbL3rFFMqrR_ej2iX:-KXlbL3rFFMqrR_ej2iY:b-h31m4q) ([raw file](https://github.com/robotlocomotion/drake/blob/4655d3057c61d3a198988f81fc61d3940f936c5d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L439)):*
> ```cpp
>   const Vector3<T> u(z_axis_W.cwiseAbs());
>   const int minAxis =
>       u[0] <= u[1] ? (u[0] <= u[2] ? 0 : 2) : (u[1] <= u[2] ? 1 : 2);
> ```

BTW, this is fine as is but Eigen does have a `minCoeff()` method that might be doing the same thing as this in a cleaner way.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 448 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlcj2BQN6V_h1w8OnT:-KXlcj2Ca32NqWgWNB9I:b4zl8v8) ([raw file](https://github.com/robotlocomotion/drake/blob/4655d3057c61d3a198988f81fc61d3940f936c5d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L448)):*
> ```cpp
>   // being tested does not actually require a valid RigidBodyTree.  So, this
>   // uninitialized tree is sufficient.
>   RigidBodyPlant<double> plant(make_unique<RigidBodyTree<double>>());
> ```

BTW, you can delete the above line and its comment now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXlczsKECoy4kHP2gPL:b-2ibfd)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Done

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 439 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlbL3rFFMqrR_ej2iX:-KXlo_cARVxYOgosT1tr:b6ppqrk) ([raw file](https://github.com/robotlocomotion/drake/blob/4655d3057c61d3a198988f81fc61d3940f936c5d/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L439)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, this is fine as is but Eigen does have a `minCoeff()` method that might be doing the same thing as this in a cleaner way.
</blockquote></details>

BTW Groovy.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 448 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-KXlcj2BQN6V_h1w8OnT:-KXloMBHnj-DFBsq1tfo:b-n9eq4d) ([raw file](https://github.com/robotlocomotion/drake/blob/4655d3057c61d3a198988f81fc61d3940f936c5d/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L448)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, you can delete the above line and its comment now.
</blockquote></details>

BTW Done  D'oh!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4272#-:-KXlpZdUPy6swklWbDMK:bb6un1j)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
@drake-jenkins-bot retest this please
